src/Core/Events/EventService.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.ObjectModel;
+using System.Linq;
+
+namespace Microsoft.Quantum.IQSharp
+{
+    /// <summary>
+    /// A typed Event, similar to EventHandler<> but without actually being a delegate
+    /// </summary>
+    /// <typeparam name="TArgs">The type of the arguments</typeparam>
+    public class Event<TArgs>
+    {
+        public TArgs Args { get; set; }
+    }
+
+    /// <summary>
+    /// A basic multi-thread implementation of a publisher-subscriber based on
+    /// the type of the event.
+    /// </summary>
+    public class EventPubSub
```

Comment:
It feels like this is an internal-facing class, since the typed wrapper is what's used directly?